### PR TITLE
Fix setting offset to a List

### PR DESCRIPTION
**Pull request by VincentLanglet**
Extracted from https://github.com/phpstan/phpstan-src/pull/3381

### CODE REVIEW

File: tests/PHPStan/Analyser/nsrt/array-is-list-offset.php [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+	/**
+	 * @param array{bool, bool, bool} $array
+	 * @param int<0, 1> $key
+	 */
+	public function test(array $array, int $key) {
+		assertType('int<0, 1>', $key);
+		assertType('true', array_is_list($array));
+
+		$array[$key] = false;
+		assertType('true', array_is_list($array));
```
Comment (VincentLanglet) [inline tests/PHPStan/Analyser/nsrt/array-is-list-offset.php:17]: Prior to this PR, it was considered as false.

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]
Hey, I specifically asked for ConstantArrayTypeBuilderTest reproduction so please add that.

**Comment by VincentLanglet** [comment]
> Hey, I specifically asked for ConstantArrayTypeBuilderTest reproduction so please add that.

Sorry for the misunderstanding, it should be better now (?)

File: tests/PHPStan/Type/Constant/ConstantArrayTypeBuilderTest.php [hunk 2]
```diff
@@ -128,4 +129,29 @@ public function testIsList(): void
...
+		$builder->setOffsetValueType(new ConstantIntegerType(0), new NullType());
+		$this->assertTrue($builder->isList());
+
+		$builder->setOffsetValueType(new ConstantIntegerType(1), new NullType());
+		$this->assertTrue($builder->isList());
+
+		$builder->setOffsetValueType(new ConstantIntegerType(2), new NullType());
+		$this->assertTrue($builder->isList());
+
+		$oneOrZero = TypeCombinator::union(
```
Comment (ondrejmirtes) [inline tests/PHPStan/Type/Constant/ConstantArrayTypeBuilderTest.php:148]: Please also write some counter-example test as well. Like when we're setting a union offset that does not exist yet, that will make the array be no longer a list. Thanks.

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]
Otherwise I couldn't find any problem with this edit 👍

**Comment by ondrejmirtes** [comment]
Thank you.

**Pull request merged**